### ISSUE TITLE
Add a hook to parseDate()

### ISSUE THREAD

**Issue by DyaGa** [contao/core#4260]
First of all I have to say that it will be hard to explain what I mean, but I will try.

So, I just found an interesting i18n issue when parsing a date.

Example:

```php
<?php
echo $this->parseDate('d F Y');
```

The result will be (examples for languages):
```
en => 29 April 2012
ru => 29 Апрель 2012
uk => 29 Квітень 2012
```

The problem is that in Ukraine or Russia we don't say `29 Апрель 2012` or `29 Квітень 2012`, we say (and the result should be):
```
ru => 29 Апреля 2012
uk => 29 Квітня 2012
```

As I see in PHP documentation about [date function](http://php.net/manual/en/function.date.php) there is no equivalent for this situation.

So, I propose you to add one more array for months (for example `MONTHS_DAY`) to language file.
Next example is for Russian language:

```php
<?php
$GLOBALS['TL_LANG']['MONTHS_DAY'][0] = 'Января';
$GLOBALS['TL_LANG']['MONTHS_DAY'][1] = 'Февраля';
$GLOBALS['TL_LANG']['MONTHS_DAY'][2] = 'Марта';
$GLOBALS['TL_LANG']['MONTHS_DAY'][3] = 'Апреля';
$GLOBALS['TL_LANG']['MONTHS_DAY'][4] = 'Мая';
$GLOBALS['TL_LANG']['MONTHS_DAY'][5] = 'Июня';
$GLOBALS['TL_LANG']['MONTHS_DAY'][6] = 'Июля';
$GLOBALS['TL_LANG']['MONTHS_DAY'][7] = 'Августа';
$GLOBALS['TL_LANG']['MONTHS_DAY'][8] = 'Сентября';
$GLOBALS['TL_LANG']['MONTHS_DAY'][9] = 'Октября';
$GLOBALS['TL_LANG']['MONTHS_DAY'][10] = 'Ноября';
$GLOBALS['TL_LANG']['MONTHS_DAY'][11] = 'Декабря';
```

And in [System->parseDate function](https://github.com/contao/core/blob/2.11.x/system/libraries/System.php#L451) make some improvements, like this:

```php
<?php
	static function parseDate($strFormat, $intTstamp=null)
	{
		$strModified = str_replace
		(
			array('l', 'D', 'd F', 'j F', 'F', 'M'),
			array('w::1', 'w::2', 'd n::5', 'j n::5', 'n::3', 'n::4'),
			$strFormat
		);
[cutted]
		$chunks = preg_split("/([0-9]{1,2}::[1-5])/", $strDate, -1, PREG_SPLIT_DELIM_CAPTURE);
[cutted]

				case 5:
					$strReturn .= $GLOBALS['TL_LANG']['MONTHS_DAY'][($index - 1)];
					break;

```

The main idea is to replace `d F` (and `j F`) with new i18n equivalent. And in this case we will have right i18n date string:

```
d F Y => 29 Апреля 2012  or  29 Марта 2012
F Y => Апрель 2012 or Март 2012
```

It's interesting to know the situation in other languages. I can only tell about Ukrainian and Russian (and till now we have a trouble with this).

**Comment by hamidabbaszadeh** [comment]
But this hook work only when there is (l, D, F and M) in $strFormat, without this it always return $strDate in line 618. I used this HOOK for parsing Persian date in front end. Is it possible to change the place of HOOK in date.php?

**Comment by leofeyer** [comment]
@hamidabbaszadeh Where would you add the hook?

**Comment by hamidabbaszadeh** [comment]
I think you must consider parse() function. By modifying parse() and adding Hook in first line with a return:
```
		// HOOK: add custom logic (see #4260)
		if (isset($GLOBALS['TL_HOOKS']['parseDate']) && is_array($GLOBALS['TL_HOOKS']['parseDate']))
		{
			foreach ($GLOBALS['TL_HOOKS']['parseDate'] as $callback)
			{
				return \System::importStatic($callback[0])->$callback[1]($strReturn, $strFormat, $intTstamp);
			}
		}
```
And everything is fine.

**Comment by leofeyer** [comment]
Fixed in b53c69d027e67752df3b80afd9803cdbf9d31e26.
